### PR TITLE
[BEAM-5959] KMS support for BigQuery

### PR DESCRIPTION
**Pull request by udim**
Adds withKmsKey() methods to reads and writes. Affects tables created
tables by: streaming writes, batched writes, query temporary tables,
table copies (for very large batched writes).

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
R: @charlesccychen 
CC: @chamikaramj

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
Run Java PostCommit

**Comment by charlesccychen** [comment]
Thanks, this LGTM.

**Comment by udim** [comment]
run seed job

**Comment by charlesccychen** [comment]
Run Java PostCommit

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
run seed job

**Comment by charlesccychen** [comment]
Can you take a look at the file presubmit tests? For example https://builds.apache.org/job/beam_PostCommit_Java_PR/51/testReport/org.apache.beam.sdk.io.gcp.bigquery/BigQueryKmsKeyIT/classMethod/

**Comment by udim** [comment]
@charlesccychen the --kmsKey test flag requires "run seed job" to be passed

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
Run Java PostCommit

**Comment by charlesccychen** [comment]
Retest this please

**Comment by charlesccychen** [comment]
Run Java PostCommit

**Comment by charlesccychen** [comment]
run seed job

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
Retest this please

**Comment by udim** [comment]
retest this please

**Comment by kennknowles** [comment]
Just checking - this is the fix for the breakage on master, yes?

**Comment by udim** [comment]
@kennknowles yes this should hopefully fix postcommits.

**Comment by udim** [comment]
Run Java PostCommit

**Comment by udim** [comment]
Run Java PreCommit

**Comment by udim** [comment]
@charlesccychen PTAL, post commit tests finally pass

**Comment by udim** [comment]
Run JavaPortabilityApi PreCommit

**Comment by charlesccychen** [comment]
Thanks!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -1534,6 +1550,10 @@ static String getExtractDestinationUri(String extractDestinationDir) {
       return toBuilder().setIgnoreUnknownValues(true).build();
     }
 
+    Write<T> withKmsKey(String kmsKey) {
```
Comment (mayansalama) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Shouldn't this method be public?

Comment (udim) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Yes, it should be.

Comment (mayansalama) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Can we reopen this as a hotfix, or should I create a new issue to track this? As far as I can tell there's no way to actually use this feature.

Comment (kennknowles) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: If you want a patch to be applied to the current LTS family (version 2.7) then please open a new ticket for that.

Comment (udim) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: @mayansalama I fixed this here: https://github.com/apache/beam/pull/8145. I should have updated you.
@kennknowles KMS support is too new to be in 2.7. :)

**Pull request merged**